### PR TITLE
Creates the validators registry

### LINKED ISSUE #79: Implement validators registry
## Problem

The yaml header is meant to be flexible and there is no imposition of any specific fields that must be there. However, there are fields that, if present, must follow certain schema. An example of this are the CSV dialect (that will be implemented in #2) and the Tabular schema (implemented in #3). It is easy to see that other applications and tools might require their own strict set of fields, eg. an measurement equipment that provides as metadata machine make and model, start and end time of the measurements, and some specific machine settings. All of those must be there if the metadata is to truly represent the data it is accompanying. 

## Proposed solution

Implement a validator registry that will run specific, known validators on the header if a given field is present. Eg. assuming #3 is implemented

```python
@register_validator(name="tabular")
class TabularSchema(...)
    ...

# and then, when reading the yaml header
for field, validator in VALIDATORS_REGISTRY.items():
    if field in yaml_header:
        yaml_header[field] = validator(field)
```
So the TabularSchema validator will run if `tabular` is a top level field in the yaml header. The same would be done when saving the data, but in that case, the validator would be dumping the data to a dict, so it can be saved.

This issue is not about implementing any specific validator, but about setting up the infrastructure for running them.

### PR DESCRIPTION
**Pull request by dalonsoa**
Additionally, run the validators when reading the header, converting the relevant fields into a pydantic object, and dump the data into dictionaries from the pydantic models when writing the data. 

Close #79

### CODE REVIEW

File: csvy/validators.py [hunk 1]
```diff
@@ -1,14 +1,91 @@
...
+        if not issubclass(cls, BaseModel):
+            raise TypeError("Validators must be subclasses of pydantic.BaseModel.")
+
+        if name in VALIDATORS_REGISTRY and not overwrite:
+            raise ValueError(f"Validator with name '{name}' already exists.")
+
+        VALIDATORS_REGISTRY[name] = cls
+        return cls
+
+    return decorator
```
Comment (alexdewar) [inline csvy/validators.py:38]: I'd usually use the `decorator` package for this to simplify the syntax a little: https://pypi.org/project/decorator/

But this is totally fine as is.

File: tests/test_write.py [hunk 2]
```diff
@@ -250,11 +250,15 @@ def test_write_data(mock_write_csv):
...
+    KNOWN_WRITERS[0].assert_called_once_with(  # type: ignore [attr-defined]
+        filename, data, comment, **csv_options
+    )
     mock_write_csv.assert_not_called()
 
     KNOWN_WRITERS.clear()
     KNOWN_WRITERS.append(MagicMock(return_value=False))
     write_data(filename, data, comment, **csv_options)
-    KNOWN_WRITERS[0].assert_called_once_with(filename, data, comment, **csv_options)
+    KNOWN_WRITERS[0].assert_called_once_with(  # type: ignore [attr-defined]
```
Comment (alexdewar) [inline tests/test_write.py:261]: You could also get `mypy` to ignore the `tests` folder if you can't be bothered with all these `type: ignore`s.

**Review by alexdewar — APPROVED** [review]
I don't know `pydantic` super well, but this LGTM!

File: tests/test_write.py (hunk 2 above)
Comment (dalonsoa) [reply tests/test_write.py:261]: It has caught a few issues, so I rather not exclude the tests and just ignore the problematic lines.

File: csvy/validators.py (hunk 1 above)
Comment (dalonsoa) [reply csvy/validators.py:38]: I don't think this works well with classes, in particular in this case where we are just registering the class using an input to the decorator itself, not the function. So skipping, for now.

**Pull request merged**